kale/loaddata/multi_domain.py

```diff
@@ -229,3 +232,257 @@ def _split_dataset_few_shot(dataset, n_fewshot, random_state=None):
     labeled_dataset = torch.utils.data.Subset(dataset, tindices)
     unlabeled_dataset = torch.utils.data.Subset(dataset, uindices)
     return labeled_dataset, unlabeled_dataset
+
+
+def domain_balanced_split(domain_labels, n_partitions, split_ratios):
+    """Get indices of random split. Samples with the same domain label will be split based on the given ratios.
+        Then the indices of different domains within the same split will be concatenated.
+
+    Args:
+        domain_labels (array-like): Labels to indicate the sample from which domain.
+        n_partitions (int): Number of partitions to split, 2 <= n_partitions <= len(split_ratios) + 1.
+        split_ratios (list): Ratios of splits to be produced, where 0 < sum(split_ratios) <= 1.
+
+    Returns:
+        [list]: Indices for different splits.
+    """
+    domains = np.unique(domain_labels)
+    subset_idx = [[] for i in range(n_partitions)]
+    for domain_label_ in domains:
+        domain_idx = np.where(domain_labels == domain_label_)[0]
+        subsets = split_by_ratios(torch.from_numpy(domain_idx), split_ratios)
+        for i in range(n_partitions):
+            subset_idx[i].append(domain_idx[subsets[i].indices])
+
+    for i in range(n_partitions):
+        subset_idx[i] = np.concatenate(subset_idx[i])
+
+    return subset_idx
+
+
+class MultiDomainImageFolder(VisionDataset):
+    """A generic data loader where the samples are arranged in this way: ::
+            root/domain_a/class_x/xxx.ext
+            root/domain_a/class_x/xxy.ext
+            ...
+            root/domain_a/class_y/xxz.ext
+            ...
+            root/domain_k/class_x/123.ext
+            root/domain_k/class_x/abc3.ext
+            ...
+            root/domain_k/class_y/asd932_.ext
+        Args:
+            root (string): Root directory path.
+            loader (callable): A function to load a sample given its path.
+            extensions (tuple[string]): A list of allowed extensions. Either extensions or is_valid_file should be
+                passed.
+            transform (callable, optional): A function/transform that takes in a sample and returns a transformed
+                version.  E.g, ``transforms.RandomCrop`` for images.
+            target_transform (callable, optional): A function/transform that takes in the target and transforms it.
+            sub_domain_set (list): A list of domain names, which should be a subset of domains in image folders.
+                Defaults to None.
+            sub_class_set (list): A list of class names, which should be a subset of classes in image folders.
+                Defaults to None.
+            is_valid_file (callable, optional): A function that takes path of a file and check if the file is a valid
+                file (used to check of corrupt files). Either extensions or is_valid_file should be passed.
```

Comment:
"used to check of corrupt files": check is a verb, *of* should be removed. "used" is not necessary, just say ("to check"/"in case of").